Feature/improve wrench computation

rjgriffin42 opened the pull request.

hrp2-14 commented:
Hi ! This project doesn't usually accept pull requests on the main branch.
If this wasn't intentionnal, you can change the base branch of this PR to devel
(No need to close it for that). Best, a bot.

cmastalli commented on CHANGELOG.md:
```diff
@@ -6,6 +6,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## [Unreleased]
 
+* Updated the rotation transpose in contact wrench cone to re-use the same object to avoid additional transpose operations.
```
Please indicate the PR as in the examples above. Also, notice that there is a conflict with other changes in the changelog.

cmastalli commented on include/crocoddyl/core/costs/cost-sum.hpp:
```diff
@@ -121,6 +121,13 @@ class CostModelSumTpl {
+   *
+   * @param[in] cost_item Cost item
+   */
+  void addCost(const boost::shared_ptr<CostItem>& cost_item);
```
1. We recently moved to std::shared_ptr, as you can see in this https://github.com/loco-3d/crocoddyl/pull/1339. 
2. I suggest defining `weight` (in the code above) as a default argument. This will require to update the Python bindings.
```suggestion

```

cmastalli replied:
The default value should be 1, of course.

cmastalli commented on include/crocoddyl/core/costs/cost-sum.hxx:
```diff
@@ -51,6 +51,26 @@ void CostModelSumTpl<Scalar>::addCost(const std::string& name,
   }
 }
 
+template <typename Scalar>
+void CostModelSumTpl<Scalar>::addCost(const boost::shared_ptr<CostItem>& cost_item)
+{
+  if (cost_item->cost->get_nu() != nu_) {
+    throw_pretty(
+        cost_item->name
+        << " cost item doesn't have the same control dimension (it should be " +
+               std::to_string(nu_) + ")");
+  }
+  costs_.insert(std::make_pair(cost_item->name, cost_item));
+  if (cost_item->active) {
+    nr_ += cost_item->cost->get_activation()->get_nr();
+    nr_total_ += cost_item->cost->get_activation()->get_nr();
+    active_set_.insert(cost_item->name);
+  } else {
+    nr_total_ += cost_item->cost->get_activation()->get_nr();
+    inactive_set_.insert(cost_item->name);
+  }
+}
+
```
From the previous comment, we need to remove this
```suggestion

```

cmastalli commented on include/crocoddyl/multibody/wrench-cone.hxx:
```diff
@@ -156,26 +159,26 @@ void WrenchConeTpl<Scalar>::update() {
     Scalar theta_i = theta * static_cast<Scalar>(i);
     Vector3s tsurf_i = Vector3s(cos(theta_i), sin(theta_i), Scalar(0.));
     Vector3s mu_nsurf = -mu * Vector3s::UnitZ();
-    A_.row(2 * i).head(3) = (mu_nsurf + tsurf_i).transpose() * R_.transpose();
+    A_.row(2 * i).head(3) = (mu_nsurf + tsurf_i).transpose() * R_transpose;
```
I have made these improvements in private, but I can discard them.
```suggestion
    A_.row(2 * i).template head<3>() = (mu_nsurf + tsurf_i).transpose() * R_transpose;
```

cmastalli commented on include/crocoddyl/multibody/wrench-cone.hxx:
```diff
@@ -299,6 +302,7 @@ void WrenchConeTpl<Scalar>::set_box(const Vector2s& box) {
 
 template <typename Scalar>
 void WrenchConeTpl<Scalar>::set_mu(const Scalar mu) {
+  mu_ = mu;
```
Nice catch! Thanks for fixing this.

This also means we're not properly unit-testing this. I would define a `test_setter` function in https://github.com/loco-3d/crocoddyl/blob/master/unittest/test_wrench_cone.cpp

cmastalli commented on include/crocoddyl/multibody/wrench-cone.hxx:
```diff
@@ -156,26 +159,26 @@ void WrenchConeTpl<Scalar>::update() {
+        (mu_nsurf - tsurf_i).transpose() * R_transpose;
   }
-  A_.row(nf_).head(3) = R_.col(2).transpose();
+  A_.row(nf_).head(3) = R_transpose.row(2);
```
```suggestion
  A_.row(nf_).template head<3>() = R_transpose.row(2);
```

cmastalli commented on include/crocoddyl/multibody/wrench-cone.hxx:
```diff
@@ -156,26 +159,26 @@ void WrenchConeTpl<Scalar>::update() {
     Vector3s mu_nsurf = -mu * Vector3s::UnitZ();
-    A_.row(2 * i).head(3) = (mu_nsurf + tsurf_i).transpose() * R_.transpose();
+    A_.row(2 * i).head(3) = (mu_nsurf + tsurf_i).transpose() * R_transpose;
     A_.row(2 * i + 1).head(3) =
```
```suggestion
    A_.row(2 * i + 1).template head<3>() =
```

cmastalli commented on include/crocoddyl/multibody/wrench-cone.hxx:
```diff
@@ -142,9 +142,12 @@ void WrenchConeTpl<Scalar>::update() {
   }
 
+  // Create a temporary object for computation
+  Matrix3s R_transpose = R_.transpose();
```
Should we allocate this data once (i.e., as a member of the structure)? Not sure if it is justified in this case. Any thought?

cmastalli reviewed (changes requested):
Thanks, @rjgriffin42!

A few, yet simple, minor requests.

rjgriffin42 commented:
@cmastalli  I updated some of the computation of the wrench cone to take advantage of some of the blocks being the same. I also updated the tests to include a setter test, as requested!

cmastalli reviewed (approved):
Thanks for pushing extra changes, @rjgriffin42!

Note that I have to fixed some compilation issues with the both header and Python bindings. Additionally, I propagated similar improvements to the friction cone.

This PR should be ready to be merged. Let's just wait if all tests are passing.
